### ISSUE TITLE
Time series doesn't exists in per-day index

### ISSUE THREAD

**Issue by lmarszal** [VictoriaMetrics/VictoriaMetrics#3309]
**Version**
v1.79.2-cluster
Can't check on the latest version; I believe the problem is data persistance (querying looks fine) and I found no way to simulate this data being ingested.

**Symptoms:**
I'm running cluster version of victoriametrics with vmagent scraping pods.

I'm running a simple query `http_request_duration_seconds_bucket{app="XXX",path="PATH",pod="XXX-6ccb9f884b-tv5c9",le="0.25"}` (sorry for masking label values). There's exactly one timeseries matching this query (and replication factor of 3). The first sample of the timeseries has 1666915196457 (2022-10-27 23:59:56.457 UTC) timestamp, the last one is 1666986071458 (2022-10-28 19:41:11.458 UTC).

I try to run a query that spans over more than one day (UTC): `/select/0/prometheus/api/v1/query_range?query=XXX&start=1666908000&end=1666994400&step=30&nocache=true&trace=true`:
- traces shows underlying vmstorage is using per-day-cache for both 2022-10-27 and 2022-10-28
  - per-day-cache for 2022-10-27 returns one matching tsid
  - per-day-cache for 2022-10-28 returns zero matching tsid
- I receive back 2362 samples first timestamp being 1666915200 (2022-10-28 00:00:00 UTC), last timestamp being 1666986060 (2022-10-28 19:41:00 UTC)

When I adjust the query to anything that doesn't span over 2022-10-27 (i.e. `start=1666918800&end=1666994400&step=30`)  I receive an empty result. Traces shows it only checks 2022-10-28 per-day cache which yields zero matching tsids.

It looks like the timeseries is present in 2022-10-27 per-day index, but not in 2022-10-28.

I also tried to use tsdb status endpoint (I believe it's using per-day cache as well). Returns results for 2022-10-27 but no results for 2022-10-28:
```
$ curl https://YYYY/select/0/prometheus/api/v1/status/tsdb -d 'match[]={__name__="http_request_duration_seconds_bucket",app="XXX",path="PATH",pod="XXX-6ccb9f884b-tv5c9",le="0.25"}' -d 'date=2022-10-27'
{"status":"success","isPartial":false,"data":{"totalSeries":3,"totalLabelValuePairs":39,"seriesCountByMetricName":[{"name":"http_request_duration_seconds_bucket","value":3}],"seriesCountByLabelName":[{"name":"__name__","value":3},{"name":"app","value":3},{"name":"code","value":3},{"name":"domain","value":3},{"name":"instance","value":3},{"name":"job","value":3},{"name":"le","value":3},{"name":"method","value":3},{"name":"path","value":3},{"name":"pod","value":3}],"seriesCountByFocusLabelValue":[],"seriesCountByLabelValuePair":[{"name":"__name__=http_request_duration_seconds_bucket","value":3},{"name":"app=XXX","value":3},{"name":"code=2xx","value":3},{"name":"domain=prod","value":3},{"name":"instance=10.71.57.23:8351","value":3},{"name":"job=XXX","value":3},{"name":"le=0.25","value":3},{"name":"method=POST","value":3},{"name":"path=PATH","value":3},{"name":"pod=XXX-6ccb9f884b-tv5c9","value":3}],"labelValueCountByLabelName":[{"name":"__name__","value":1},{"name":"app","value":1},{"name":"code","value":1},{"name":"domain","value":1},{"name":"instance","value":1},{"name":"job","value":1},{"name":"le","value":1},{"name":"method","value":1},{"name":"path","value":1},{"name":"pod","value":1}]}}
```
```
$ curl https://victoriametrics.eks-prod-ca.k8s.remitly.com/select/0/prometheus/api/v1/status/tsdb -d 'match[]={__name__="http_request_duration_seconds_bucket",app="XXX",path="PATH",pod="XXX-6ccb9f884b-tv5c9",le="0.25"}' -d 'date=2022-10-28'
{"status":"success","isPartial":false,"data":{"totalSeries":0,"totalLabelValuePairs":0,"seriesCountByMetricName":[],"seriesCountByLabelName":[],"seriesCountByFocusLabelValue":[],"seriesCountByLabelValuePair":[],"labelValueCountByLabelName":[]}}
```

Also tried to use `/api/v1/export` - works perfectly fine:
```
$ curl https://YYY/select/0/prometheus/api/v1/export -d 'match[]={__name__="http_request_duration_seconds_bucket",app="XXX",path="PATH",pod="XXX-6ccb9f884b-tv5c9",le="0.25"}'
{"metric":{"__name__":"http_request_duration_seconds_bucket","job":"XXX","instance":"10.71.57.23:8351","pod":"XXX-6ccb9f884b-tv5c9","app":"XXX","code":"2xx","domain":"prod","le":"0.25","method":"POST","path":"PATH","prometheus":"victoriametrics/apps","squad":"ZZZ","version":"VVV"},"values":[TOO LONG FOR GITHUB],"timestamps":[TOO LONG FOR GITHUB]}
```

Just an observation - first timestamp is super close to midnight, when the actual rotation of the indexes happen.

Might be related to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3305 as symptoms looks similar there.

This is way out of my depth, is anyone available to help bottom up this issue?

**Comment by lmarszal** [comment]
Ok - I reproduced the issue.
<img width="620" alt="image" src="https://user-images.githubusercontent.com/26798897/200020329-f9c29c80-d308-40fb-b15f-baf6eead5752.png">
The problem is that:
1. `s.currHourMetricIDs` may contain metricIDs that appeared in the previous hour ([ref](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/94bd49402ec96a720b9807a188865e204ad73342/lib/storage/storage.go#L2324-L2356))
2. `updatePerDateData` assumes, that if `currHourMetricIDs` contains metricId, it doesn't need to create per-day index entries ([ref](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/94bd49402ec96a720b9807a188865e204ad73342/lib/storage/storage.go#L1984-L2004))

If you're unlucky and your first datapoint arrives between last execution of `updateCurrHourMetricIDs` on a given data and midnight and your second datapoint arrives after first execution of `updateCurrHourMetricIDs` on a given day, the arrival of 2nd datapoint won't create per-day index (even though it should).  
And, since metric is already registered in a current or previous hour cache for a given day ([ref](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/94bd49402ec96a720b9807a188865e204ad73342/lib/storage/storage.go#L2005-L2008)), the error will never self-heal until the day passes.

Please see this code for reproduction: https://github.com/lmarszal/VictoriaMetrics/tree/issue-3309-reproduction  

You need to have `go`, `bash`, `curl` and `jq` installed. You simply run `repro.sh` to reproduce the issue. There are couple changes to the main branch to reproduce it (see [compare](https://github.com/VictoriaMetrics/VictoriaMetrics/compare/master...lmarszal:VictoriaMetrics:issue-3309-reproduction?expand=1)):
- `lib/fasttime/fasttime.go` is equipped with a "time travel" mock, that simulates a situation where victoria starts 15s before midnight
- `currHourMetricIDsUpdateInterval` is changed from `10s` to `30s` so that it's easier to exec a request between `updateCurrHourMetricIDs` runs
- some extra logging (not for production - just to see what happens during the repro)
The `repro.sh` script build victoria, starts it, adds first datapoint, waits, adds 2nd datapoint, waits and queries victoria to see the traces that proves that only one per-day index contains the timeseries.

I'll have some time to work on the fix (and PR) on the week of Nov 14th - unless repository owners will fix it sooner.

**Comment by f41gh7** [comment]
Thanks for detailed bug report, will fix it soon.

**Comment by lmarszal** [comment]
> Thanks for detailed bug report, will fix it soon.

Thanks for the reply @f41gh7 . Please see my fix proposal: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/3320

**Comment by valyala** [comment]
@lmarszal , thanks for the bugfix! It has been included in the commit 790768f20b7ed83af315fcc37ee15d3fd11b33e3 . This commit will be included in the next release of VictoriaMetrics. See also the follow-up commit at daa70e6560b23e931c56fe38614ca745c727f514 . Re-opening the issue until the next release.

**Comment by valyala** [comment]
The bug should be fixed starting from [VictoriaMetrics v1.83.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue then.

**Comment by valyala** [comment]
FYI, the bugfix for this issue is also included in [v1.79.5 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.79.5).

**Comment by valyala** [comment]
It has been appeared that the `v1.79.5` release contains incomplete fix for the issue, which could lead to high CPU usage at zero UTC hour every day. The issue has been fixed in [v1.79.10 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.79.10).
